src/platform_impl/android/mod.rs

```diff
@@ -1012,46 +1003,49 @@ impl MonitorHandle {
         self.app.config().density().map(|dpi| dpi as f64 / 160.0).unwrap_or(1.0)
     }
 
-    pub fn refresh_rate_millihertz(&self) -> Option<u32> {
-        // FIXME no way to get real refresh rate for now.
-        None
-    }
-
-    pub fn video_modes(&self) -> impl Iterator<Item = VideoModeHandle> {
-        let size = self.size().into();
-        // FIXME this is not the real refresh rate
-        // (it is guaranteed to support 32 bit color though)
-        std::iter::once(VideoModeHandle {
-            size,
+    pub fn current_video_mode(&self) -> Option<VideoModeHandle> {
+        Some(VideoModeHandle {
+            size: screen_size(&self.app),
+            // FIXME: it is guaranteed to support 32 bit color though
             bit_depth: 32,
```

Comment:
They seem to have [deprecated the ability to get the pixel format](https://developer.android.com/reference/android/view/Display#getPixelFormat()) long ago, unfortunately.

There's no point querying the format of the `Surface` since:
1. There won't be a `Surface` until the `Resumed` event;
2. The format typically starts out with `RGB565` for some arcane reason;
3. The user can freely change this pixel format and the compositor will Deal With It™.

Maybe this needs to be `Option` as well?

Reply:
> 1. There won't be a `Surface` until the `Resumed` event;

Why would that make it pointless? Do you mean because when you have a surface you use your graphics API to determine your pixel format anyway?

In that case the API would be useless on all platforms wouldn't it?
Personally I don't know who would use that, so I would actually be in favor of removing it wholesale if we come to a consensus.

> Maybe this needs to be `Option` as well?

I think we should change it to `None` if its not supported.

My assumption was always that the main use-case here is to check for HDR support, which Android would support with [`android.view.Display.getSupportedHdrTypes()`](https://developer.android.com/reference/android/view/Display.Mode#getSupportedHdrTypes()). But that API should probably look different and again, not useful to handle in Winit. E.g., we discussed querying pixel format support in Softbuffer already https://github.com/rust-windowing/softbuffer/issues/98 and Wgpu obviously already supports that.

---

If you don't mind I would leave it like this here and make a follow up PR discussing any further changes.
WDYT in any case about all this?

Reply:
> > 1. There won't be a `Surface` until the `Resumed` event;
> 
> Why would that make it pointless? Do you mean because when you have a surface you use your graphics API to determine your pixel format anyway?

Pointless _on Android_, since I think you can query the video mode (and most apps probably will?) before a "surface" appears?

The pixel format of a surface can and will almost always be disjoint from the bit depth of the video mode on a physical panel/connection because of compositors (though it's probably beneficial to keep them in sync depending on what the scanout hardware is capable of).

> In that case the API would be useless on all platforms wouldn't it? Personally I don't know who would use that, so I would actually be in favor of removing it wholesale if we come to a consensus.

I think all platforms with a list of monitors and their current mode can perfectly benefit from this API. A pixel format may be more helpful over a plain bit depth though, but that's bikeshedding at this point.

> > Maybe this needs to be `Option` as well?
> 
> I think we should change it to `None` if its not supported.

Exactly: if this function can return `Option`, its value will always be `None` on Android.

> My assumption was always that the main use-case here is to check for HDR support, which Android would support with [`android.view.Display.getSupportedHdrTypes()`](https://developer.android.com/reference/android/view/Display.Mode#getSupportedHdrTypes()). 

You can't do that with bit-depth alone, e.g. RGBA8888 and R10G10B10A2 both use 32 bits. Both could be used for a HDR _color space_ but the former likely has visible banding.

> But that API should probably look different and again, not useful to handle in Winit. 

Yeah, HDR is much, _much_ more complicated than that. Better leave it to a users' Vulkan/Dxgi/(E)GL implementation.

Though if a platform provides additional HDR-related details for the current/negotiated _mode_ for a monitor via its native API, it _might_ be nice to expose that on the mode handle.
(things like HDR metadata, VRR, color space from the mode; you have that in Windows and DRM at least)

> E.g., we discussed querying pixel format support in Softbuffer already https://github.com/rust-windowing/softbuffer/issues/98 and Wgpu obviously already supports that.

Seems like that should be disconnected from winit, yes.

> If you don't mind I would leave it like this here and make a follow up PR discussing any further changes. WDYT in any case about all this?

:+1:

Reply:
> The pixel format of a surface can and will almost always be disjoint from the bit depth of the video mode on a physical panel/connection because of compositors (though it's probably beneficial to keep them in sync depending on what the scanout hardware is capable of).

...

> > In that case the API would be useless on all platforms wouldn't it? Personally I don't know who would use that, so I would actually be in favor of removing it wholesale if we come to a consensus.
> 
> I think all platforms with a list of monitors and their current mode can perfectly benefit from this API. A pixel format may be more helpful over a plain bit depth though, but that's bikeshedding at this point.

Well that would still only make sense if the bit depth has anything to do with the pixel format when you start drawing. Your earlier statement seems to say that this isn't a guarantee?

> > > Maybe this needs to be `Option` as well?
> > 
> > 
> > I think we should change it to `None` if its not supported.
> 
> Exactly: if this function can return `Option`, its value will always be `None` on Android.

Yeah, that's fine. Other platforms don't support querying it as well.

---

It seems this discussion largely boils down to bit depth vs pixel format. In Winit, the bit depth is only useful when you *choose* a video mode handle, which you can via `Window::set_fullscreen()` (not on Android though). But this all implies that the bit depth has some connection to the pixel format, obviously the bit depth doesn't contain all the information of a pixel format, but it does make a difference. E.g. on a 6-bit video mode you can't use 8-bit color pixel format.

In reality of course, the choice is usually between 8-bit (99.9%) and 10-bit per color. No clue how HDR plays a role here, but as we said this would require a different API.

Seems all pointless on Android though if you can't use that information in `Window::set_fullscreen()`, so this basically boils down to use just exposing device data.

---

Conclusion: make it an `Option`.
(will do in a follow-up)

Reply:
I've done it in this PR after all.

Reply:
> It seems this discussion largely boils down to bit depth vs pixel format.

Can't help when discussing how to implement this function, we look at where to pull the necessary information from, and eventually what the API even means and how it is useful. Let me have a look at `Window::set_fullscreen()` and how it utilizes the `bit_depth()` specifically.

> In reality of course, the choice is usually between 8-bit (99.9%) and 10-bit per color.

Again, it's hard to tell those apart if you have `R10G10B10A2` versus `R8G8B8A8` or `R8G8B8X8` which are both 32-bits. We'd need to return the _per-channel_ bit depth in that case?

Reply:
> > In reality of course, the choice is usually between 8-bit (99.9%) and 10-bit per color.
> 
> Again, it's hard to tell those apart if you have `R10G10B10A2` versus `R8G8B8A8` or `R8G8B8X8` which are both 32-bits. We'd need to return the _per-channel_ bit depth in that case?

It depends if those are part of video modes. If they are pixel formats then I don't think the are relevant to Winit.

Reply:
The only two relevant platforms for this function appear to be Windows and X11, where Windows has it defined here:

https://learn.microsoft.com/en-us/windows/win32/api/wingdi/ns-wingdi-devmodea

Seems to stem from an era where we had less than 32 bits per pixel to define a color (or palette?) and not even a single clue about the pixel _format_ (and related color space).

Let's postpone this discussion now that we can return `None` everywhere this function makes no sense.